### DISCUSSION TITLE
all RTL grid based components shifts weirdly

### DISCUSSION THREAD

**Discussion by benllshua** [saadeghi/daisyui#2211] [Q&A] [answered]
Hi, I wonder if anyone has encountered this issue with RTL 
All the grid-based component order shifts ( all connectors get moved one item forward or something) 

![image](https://github.com/saadeghi/daisyui/assets/47493928/46024653-a424-45b5-9e1a-5abaeabad8eb)
![image](https://github.com/saadeghi/daisyui/assets/47493928/bf7c1b7a-d555-43d9-a07f-8becbbc868e2)

**Comment by benllshua** [comment]
@saadeghi would love some help :)

**Comment by saadeghi** [comment]
- Make sure you have `dir=rtl` on your root element (`<html>` tag)  
- Make sure you have `tailwindcss-flip` plugin
- Make sure you have `rtl: true` config

These are all mentioned on the docs: https://daisyui.com/docs/config/

https://play.tailwindcss.com/tdLN6O1OlJ

Let me know if you have any questions.

**Reply by benllshua** [reply]
hi @saadeghi  pay attension that this works only if your whole app is rtl or ltr, for example your example get shifts weirdly when dir=ltr...

I want to dynamically change ltr to rtl while in app when language changes... and I dont want to config in tailwind.config as rtl or ltr...


![image](https://github.com/saadeghi/daisyui/assets/47493928/91ffaa4e-b0ac-4e6c-84d4-0515e4ed0531)


![image](https://github.com/saadeghi/daisyui/assets/47493928/3e9f6ea1-b05d-47f0-9265-f71be7da1dd7)


this is from the example you sent... when the rtl: bool in the config dont match the dir="rtl"|"ltr" prop in the root element

**Reply by benllshua** [reply]
I would suggest remove the rtl: bool from the config and set css according to root element dir prop dynamically

**Reply by saadeghi** [reply, marked as answer]
@benllshua Yes. Everything should be at build time.  
All the 3 things I mentioned should be done to generate a RTL file and all those 3 things should be undone for LTR mode.  

Currently this is the only practical way because even without daisyUI, people use a lot of single-direction utility classes like `mr-4` `pl-2` `right-20` etc.  
These things cannot be rotated at run time so what we're doing here is rotating the whole CSS file at build time using `tailwindcss-flip` plugin and then we apply specific style for daisyUI styles that need specific styles for RTL.

Having 1 CSS file that includes both RTL and LTR styles is not efficient either. Usually users choose a direction (language) once and they keep using the site with the same direction forever so why give all the unused LTR style to a RTL users and all the unused RTL styles to the LTR users?

If you're site is bidirectional, you can generate 2 CSS files at build time and switch them based what user picks.

Here's an example setup of having 2 `tailwind.config.js` files to generate 2 CSS files:
https://github.com/saadeghi/daisyui-bidirectional-builds

I'm planning to use logical properties so I won't need that package for rotating margins, paddings, etc. But some styles like box shadows will be a challenge that way.  
I'm working on it and hopefully I will find a way to make it work without config change for the next major version of daisyUI.

Let me know if you have any questions

**Reply by benllshua** [reply]
thank you so much for the detailed answer, I understand now the performence vs usebillity better now. Ill try the approach you gave!
